openhands/events/observation/error.py

```diff
@@ -12,6 +12,7 @@ class ErrorObservation(Observation):
     E.g., Linter error after editing a file.
     """
 
+    fatal: bool = False
```

Comment:
The reason I thought adding a FatalErrorObservation is a good thing, is that we treat ErrorObs too much the same, when there are essential differences, such as: 
- some should be in agent's history, some not;
- some should have a user-friendly message, some should have an LLM-friendly content.

In this case, it seems we can also add:
- some should stop the controller, some should not.

Not distinguishing them well right now, is IMHO a pretty big part of what is confusing in our current error handling. The alternative might be to go back to ErrorObs are in one category, and exceptions in the other... I guess.

Reply:
Hmm. I do think it's important to distinguish them (hence the `fatal=True`) but we end up having to check `isinstance(ErrorObs) or isinstance(FatalErrorObs)`, and it seems ripe for someone forgetting to handle one or the other, or for the list of error types to grow indefinitely

In general I want to pare down the number of unique things that could end up in the EventStream, because as we add long-term state, we're going to have to ensure old EventStreams still deserialize properly, and the smaller the footprint the better. 

> some should stop the controller, some should not.
This seems like the only thing `fatal` should really distinguish. I.e. if you hit a `fatal` Error, the event stream stops until there's a user interaction. I see what you mean about one being user-facing, and the other being LLM-facing though...

Interestingly, as I look through a lot of the fatal ones, they seem like things the agent should be able to recover from (e.g. submitting a non-existent action)

Will think more about this

Reply:
Aha, yes, good points. Re: deserialization, I just hit something similar.

About this bit: 
> I see what you mean about one being user-facing, and the other being LLM-facing though...

What do you think about adding a "cause" or "error" attr to the ErrorObs, with the underlying exception string, at least, if not more of the underlying exception? (We're already using a "cause" attribute, maybe we could still rename that to "parent", but yep deserialization)

So that the agent has an ErrorObs with that info. I seem to recall having tried it, though, but I didn't like the result, because we may want to be a bit careful to make it very easy for developers to use in all the places where an ErrorObs is created or where that underlying exception is handled. Otherwise it will be forgotten, as you just said. 😅 But that was a while ago, maybe it's worth another look if there is some reasonable way to do it?

Reply:
Actually, while the `isinstance()` issue is a huge red flag, that could be taken care of by subclassing...